### PR TITLE
[labs/analyzer] Add support for analyzing function declarations

### PR DESCRIPTION
**Pull request by kevinpschaaf**

### CODE REVIEW

File: packages/labs/analyzer/src/test/javascript/functions_test.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    assert.equal(fn.name, `function2`);
+    assert.equal(fn.summary, `Function 2 summary\nwith wraparound`);
+    assert.equal(fn.description, `Function 2 description\nwith wraparound`);
+    assert.equal(fn.parameters?.length, 3);
+    assert.equal(fn.parameters?.[0].name, 'a');
+    assert.equal(fn.parameters?.[0].description, 'Param a description');
+    assert.equal(fn.parameters?.[0].summary, undefined);
+    assert.equal(fn.parameters?.[0].type?.text, 'string');
+    assert.equal(fn.parameters?.[0].default, undefined);
+    assert.equal(fn.parameters?.[0].rest, false);
```
Comment (justinfagnani) [inline packages/labs/analyzer/src/test/javascript/functions_test.ts:78]: Uvu's `equal()` is a deep equal... can you use that for better readability?

**Review by justinfagnani — APPROVED** [review]

File: packages/labs/analyzer/src/lib/javascript/functions.ts [hunk 2]
```diff
@@ -12,9 +12,58 @@
...
+    );
+  }
+  return name;
+};
+
+export const getFunctionDeclarationInfo = (
+  declaration: ts.FunctionDeclaration,
+  analyzer: AnalyzerInterface
+): DeclarationInfo => {
+  const name = getFunctionDeclarationName(declaration);
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/javascript/functions.ts:48]: `getFunctionLikeInfo()` is called here and in `ClassMethod`

File: packages/labs/analyzer/src/lib/javascript/functions.ts [hunk 3]
```diff
@@ -12,9 +12,58 @@
...
+  return {
+    name,
+    factory: () => getFunctionDeclaration(declaration, name, analyzer),
+    isExport: hasExportModifier(declaration),
+  };
+};
+
+export const getFunctionDeclaration = (
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/lib/javascript/functions.ts:56]: Done

File: packages/labs/analyzer/src/test/javascript/functions_test.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+      ctx.packagePath = packagePath;
+      ctx.analyzer = analyzer;
+      ctx.module = module;
+    } catch (error) {
+      // Uvu has a bug where it silently ignores failures in before and after,
+      // see https://github.com/lukeed/uvu/issues/191.
+      console.error('uvu before error', error);
+      process.exit(1);
+    }
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/test/javascript/functions_test.ts:49]: Will separate into follow-up PR.

File: packages/labs/analyzer/src/test/javascript/functions_test.ts (hunk 1 above)
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/test/javascript/functions_test.ts:78]: It seems to do pretty bad things with classes, so gonna avoid for now.

File: packages/labs/analyzer/src/test/javascript/functions_test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/labs/analyzer/src/test/javascript/functions_test.ts:49]: https://github.com/lit/lit/pull/3661/files

**Pull request merged**